### PR TITLE
tbi(@sap-ux/odata-service-inquirer): Adds new module

### PR DESCRIPTION
**Pull request by IainSAP**
First implementation of https://github.com/SAP/open-ux-tools/issues/1864

Note this is not yet fully featured and currently will only support 2 datasources, local CAP projects and metadata (edmx) file upload. Further datasource support will be added shortly (odata service urls, SAP systems,  SAP Business Hub,  BTP Project Specific Destinations).

- Adds new module `@sap-ux/odata-service-inquirer`
- Adds datasource support for `DatasourceType.cap_project`, `DatasourceType.metadata_file` and `DatasourceType.none`
- Not-yet-implemented selections will return a warning message and log
- Consolidates re-use functions (prompt helpers) to `@sap-ux/inquirer-common`
- Adds tests to cover new functionality

**todo:**

- Add missing tsdocs
- Add sonar config

### CODE REVIEW

File: packages/odata-service-inquirer/src/prompts/prompts.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+ */
+function getDatasourceTypeQuestion(options?: DatasourceTypePromptOptions): YUIQuestion<OdataServiceAnswers> {
+    const choices = getDatasourceTypeChoices(options);
+    return {
+        type: 'list',
+        name: promptNames.datasourceType,
+        guiOptions: {
+            breadcrumb: true
+        },
+        default: options?.default ?? -1,
```
Comment (cianmSAP) [inline packages/odata-service-inquirer/src/prompts/prompts.ts:46]: Curious as to what -1 does?

File: packages/odata-service-inquirer/src/error-handler/error-handler.ts [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+    [ERROR_TYPE.BAD_GATEWAY]: [/502/],
+    [ERROR_TYPE.DESTINATION_BAD_GATEWAY_503]: [],
+    [ERROR_TYPE.DESTINATION_UNAVAILABLE]: [],
+    [ERROR_TYPE.DESTINATION_NOT_FOUND]: [],
+    [ERROR_TYPE.DESTINATION_MISCONFIGURED]: [],
+    [ERROR_TYPE.NO_V2_SERVICES]: [],
+    [ERROR_TYPE.NO_V4_SERVICES]: []
+};
+
+// Maps errors to end-user messages
```
Comment (longieirl) [inline packages/odata-service-inquirer/src/error-handler/error-handler.ts:89]: Two types of method comments here.

File: packages/odata-service-inquirer/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,282 @@
+import type { IValidationLink } from '@sap-devx/yeoman-ui-types';
+import type { YUIQuestion } from '@sap-ux/inquirer-common';
+import type { OdataVersion, ServiceType } from '@sap-ux/odata-service-writer';
+import type { CapCustomPaths } from '@sap-ux/project-access';
+import type { ListChoiceOptions } from 'inquirer';
+import type { AutocompleteQuestionOptions } from 'inquirer-autocomplete-prompt';
+
+export enum DatasourceType {
+    sap_system = 'sap_system',
```
Comment (cianmSAP) [inline packages/odata-service-inquirer/src/types.ts:9]: Is ther not inconsistencies with enums with camel case vs underscores ?

File: packages/odata-service-inquirer/src/utils/index.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    telemetryClient = toolsSuiteTelemetryClient;
+}
+
+/**
+ * Send telemetry event.
+ *
+ * @param eventName the name of the telemetry event
+ * @param telemetryData the telemetry values to report
+ */
+export function sendTelemetryEvent(eventName: string, telemetryData: TelemetryProperties): void {
```
Comment (cianmSAP) [inline packages/odata-service-inquirer/src/utils/index.ts:39]: Probably stating the obvious but we'll have to move this somewhere?

File: packages/odata-service-inquirer/src/translations/odata-service-inquirer.i18n.json [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            "capProjectName": "Local CAP Project",
+            "metadataFileName": "Metadata File",
+            "noneName": "None",
+            "projectSpecificDestChoiceText": "Connect to a $t{prompts.datasourceType.projectSpecificDestName}",
+            "businessHubChoiceText": "Connect to $t(prompts.datasourceType.businessHubName)",
+            "sapSystemChoiceText": "Connect to a $t(prompts.datasourceType.sapSystemName)",
+            "odataServiceUrlChoiceText": "Connect to an $t(prompts.datasourceType.odataServiceUrlName)",
+            "capProjectChoiceText": "Use a $t(prompts.datasourceType.capProjectName)",
+            "metadataFileChoiceText": "Upload a $t(prompts.datasourceType.metadataFileName)",
+            "notYetImplemenetdWarningMessage": "The selected data source type: {{ datasourceType }} is not yet implemented."
```
Comment (cianmSAP) [inline packages/odata-service-inquirer/src/translations/odata-service-inquirer.i18n.json:18]:
```suggestion
            "notYetImplementedWarningMessage": "The selected data source type: {{ datasourceType }} is not yet implemented."
```

**Review by cianmSAP — COMMENTED** [review]
Looks good, a few small comments

File: packages/odata-service-inquirer/src/prompts/validators.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * Validates the metadata file, returning an error string, if not valid, or an object with the valid metadata and version.
+ *
+ * @param path
+ * @param odataVersion
+ * @returns
+ */
+
+export const validateMetadataFile = async (
```
Comment (cianmSAP) [inline packages/odata-service-inquirer/src/prompts/validators.ts:52]: The `cap-project` directory has a `validators.ts` with the associated validator for cap prompt, I'm wondernig why this isn't within the `metadat-file` dir?

File: packages/odata-service-inquirer/src/prompts/prompts.ts (hunk 1 above)
Comment (IainSAP) [reply packages/odata-service-inquirer/src/prompts/prompts.ts:46]: None selected so you get the message `Please choose...`

File: packages/odata-service-inquirer/src/utils/index.ts (hunk 4 above)
Comment (IainSAP) [reply packages/odata-service-inquirer/src/utils/index.ts:40]: Not currently used anywhere else so Im going to leave it for now.

File: packages/odata-service-inquirer/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    return answers;
+}
+
+export {
+    DatasourceType,
+    OdataVersion,
+    getPrompts,
+    prompt,
+    promptNames,
+    // temp exports, remove once development is done
```
Comment (cianmSAP) [inline packages/odata-service-inquirer/src/index.ts:80]: Are these to be removed?

**Review by cianmSAP — DISMISSED** [review]
Comments addressed
Code updates to `inquirer-common` and consuming modules looks good
New module added
Coverage above 90%
Changeset 
✅

**Review by longieirl — APPROVED** [review]
Changes look good.
Good test coverage.
A number of TODO's outstanding in odata-service-inquirer.
Clear and concise comments.

**Review by cianmSAP — APPROVED** [review]
Re-approving

File: packages/odata-service-inquirer/README.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+`getPrompts` is provided to allow consumers to access prompts. There may be cases where these can be transformed to support other prompting frameworks. Most prompt configuration is possible via `OdataServiceInquirerOptions` and calling `prompt`. 
+
+Configurability of prompts is entirely controlled using the `OdataServiceInquirerOptions` parameter. 
+
+See [Inquirer.js](https://www.npmjs.com/package/inquirer) for valid default properties.
+
+### Inquirer usage example:
+
+**TODO: REPLACE EXAMPLE**
```
Comment (IainSAP) [inline packages/odata-service-inquirer/README.md:27]: Please update

File: packages/odata-service-inquirer/test/unit/prompts/cap-project/cap-helper.test.ts [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+        // Error handler should be called with custom error message
+        expect(errorHandlerSpy).toHaveBeenCalledWith(
+            t('errors.cannotReadCapServiceMetadata', { serviceName: 'CatalogService' })
+        );
+        // Logger should be called with original error, this is unreliable as the underlying string value may be changed by CDS
+        expect(logErrorSpy).toHaveBeenCalledWith(
+            expect.objectContaining({
+                message: expect.stringContaining(" Module '@sap/cds' not installed in project")
+            } as Error)
+        );
```
Comment (devinea) [inline packages/odata-service-inquirer/test/unit/prompts/cap-project/cap-helper.test.ts:192]: This is failing locally as `@sap/cds-dk` is installed globally. Could the global cds loading be mocked for the test?
`Error: Error while reading CAP service metadata. Path: '/SAPDevelop/open-ux-tools/packages/odata-service-inquirer/test/unit/prompts/cap-project/fixtures/bookshop', service uri: 'odata/v4/incident/', error: 'Error: The @sap/cds major version (>=3.33.1) specified in your CAP project is different to the @sap/cds version you have installed globally (7.4.1). Please run 'npm install' on your CAP project to ensure that the correct CDS version is loaded.'}`

File: packages/odata-service-inquirer/package.json [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        "i18next": "23.5.1",
+        "os-name": "4.0.1"
+    },
+    "devDependencies": {
+        "@sap-ux/odata-service-writer": "workspace:*",
+        "@sap-ux/feature-toggle": "workspace:*",
+        "@sap-devx/yeoman-ui-types": "1.14.4",
+        "@types/inquirer-autocomplete-prompt": "2.0.1",
+        "@types/inquirer": "8.2.6",
+        "@types/lodash": "4.14.178",
```
Comment (devinea) [inline packages/odata-service-inquirer/package.json:48]:
```suggestion
        "@types/lodash": "4.14.202",
```

File: .changeset/hip-ants-clap.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+---
+'@sap-ux/odata-service-inquirer': patch
```
Comment (devinea) [inline .changeset/hip-ants-clap.md:2]: Please add other changed modules here

**Review by devinea — COMMENTED** [review]
Code changes are clear and well structured. 
New module set-up looks good.

A few small comments.

File: packages/odata-service-inquirer/README.md [hunk 12]
```diff
@@ -0,0 +1,87 @@
+# @sap-ux/odata-service-inquirer
+
+Provides Inquirer based end-user prompting to allow selection of a service from multiple data source types. This invloves aquiring a connection to backend systems and retrieving edmx metadata for services provided by the catalog, from a local file or CAP project.
```
Comment (unseen1980) [inline packages/odata-service-inquirer/README.md:3]: ```involves```

Comment (unseen1980) [inline packages/odata-service-inquirer/README.md:3]: ```acquiring```

File: packages/odata-service-inquirer/src/prompts/datasources/cap-project/cap-helpers.ts [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+ *
+ * @param paths - The paths used to search for CAP projects
+ * @returns The CAP project prompt choices
+ */
+export async function getCapProjectChoices(paths: string[]): Promise<CapProjectChoice[]> {
+    const { capProjectPaths, folderCounts } = await getCapProjectPaths(paths);
+
+    const capChoices: CapProjectChoice[] = [];
+
+    for await (const capProjectPath of capProjectPaths) {
```
Comment (unseen1980) [inline packages/odata-service-inquirer/src/prompts/datasources/cap-project/cap-helpers.ts:56]: I would suggest Promise.all() for better performace. The getCapCustomPaths will be called in parallel for each path. Something like this:

```
export async function getCapProjectChoices(paths: string[]): Promise<CapProjectChoice[]> {
    const { capProjectPaths, folderCounts } = await getCapProjectPaths(paths);

    const capChoicesPromises = capProjectPaths.map(async (capProjectPath) => {
        const customCapPaths = await getCapCustomPaths(capProjectPath.path);
        const folderCount = folderCounts.get(capProjectPath.folderName) ?? 1;
        return {
            name: `${capProjectPath.folderName}${folderCount > 1 ? ' (' + capProjectPath.path + ')' : ''}`,
            value: Object.assign(capProjectPath, customCapPaths)
        };
    });

    const capChoices = await Promise.all(capChoicesPromises);

    return [
        ...capChoices,
        {
            name: t('prompts.capProject.enterCapPathChoiceName'),
            value: enterCapPathChoiceValue
        }
    ];
}
```

File: packages/odata-service-inquirer/src/translations/odata-service-inquirer.i18n.json [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        "datasourceType": {
+            "message": "Data source",
+            "projectSpecificDestName": "Project Specific Destination",
+            "businessHubName": "SAP Business Accelerator Hub",
+            "sapSystemName": "SAP System",
+            "odataServiceUrlName": "OData Service Url",
+            "capProjectName": "Local CAP Project",
+            "metadataFileName": "Metadata File",
+            "noneName": "None",
+            "projectSpecificDestChoiceText": "Connect to a $t{prompts.datasourceType.projectSpecificDestName}",
```
Comment (unseen1980) [inline packages/odata-service-inquirer/src/translations/odata-service-inquirer.i18n.json:12]: Does it work both with ```$t{}``` and ```$t()``` ?

File: packages/odata-service-inquirer/src/prompts/datasources/cap-project/cap-helpers.ts (hunk 13 above)
Comment (IainSAP) [reply packages/odata-service-inquirer/src/prompts/datasources/cap-project/cap-helpers.ts:56]: Not sure, would need to do some analysis to see if there's any appreciable difference.

File: packages/odata-service-inquirer/src/translations/odata-service-inquirer.i18n.json (hunk 14 above)
Comment (IainSAP) [reply packages/odata-service-inquirer/src/translations/odata-service-inquirer.i18n.json:12]: Thx, no it doesnt work.

**Comment by IainSAP** [comment]
Back to draft, partial integration of these prompts with consuming code prompts needs some additional work to be able to pass non-answer data (derived answers) where `getPrompts` only is used.

File: .changeset/hip-ants-clap.md (hunk 11 above)
Comment (IainSAP) [reply .changeset/hip-ants-clap.md:2]: Done

File: packages/odata-service-inquirer/test/unit/prompts/cap-project/cap-helper.test.ts (hunk 9 above)
Comment (IainSAP) [reply packages/odata-service-inquirer/test/unit/prompts/cap-project/cap-helper.test.ts]: Removed test as unnecessarily testing another modules unit and mocking `require.resolve` seems like too much effort and should be done in `project-access` `module-loader` unit tests.

**Comment by IainSAP** [comment]
Re-opening for re-review thx

**Review by devinea — APPROVED** [review]
Thanks for addressing comments.

File: packages/odata-service-inquirer/src/prompts/datasources/metadata-file/validators.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+      }
+    | string
+    | boolean
+> => {
+    if (!path) {
+        return false;
+    }
+
+    try {
+        const metadata = await readFile(path, 'utf-8');
```
Comment (longieirl) [inline packages/odata-service-inquirer/src/prompts/datasources/metadata-file/validators.ts:30]: File will exist before loading so no dependency on `this.fs` from the generator, adding this note incase I review it again and forget!

**Review by longieirl — APPROVED** [review]
Reviewed all the code only.
Changes look good.
Good test coverage.
Will be tested over the next few weeks as its consumed back into toolsuite.

**Review by devinea — APPROVED** [review]
Re-approving.

**Pull request merged**